src/v/kafka/server/replicated_partition.cc

```diff
@@ -109,42 +109,15 @@ replicated_partition::aborted_transactions(
   model::offset last,
   ss::lw_shared_ptr<const storage::offset_translator_state> ot_state) {
     vassert(ot_state, "ntp {}: offset translator state must be present", ntp());
-
-    // Note: here we expect that local _partition contains aborted transaction
-    // ids for both local and remote offset ranges. This is true as long as
-    // rm_stm state has not been reset (for example when there is a partition
-    // transfer or when a stale replica recovers its log from beyond the log
-    // eviction point). See
-    // https://github.com/redpanda-data/redpanda/issues/3001
-
     auto base_rp = ot_state->to_log_offset(base);
     auto last_rp = ot_state->to_log_offset(last);
-    auto source = co_await _partition->aborted_transactions(base_rp, last_rp);
-
-    // We trim beginning of aborted ranges to `trim_at` because we don't have
-    // offset translation info for earlier offsets.
-    model::offset trim_at;
-    if (base_rp >= _partition->start_offset()) {
-        // Local fetch. Trim to start of the log - it is safe because clients
-        // can't read earlier offsets.
-        trim_at = _partition->start_offset();
-    } else {
-        // Fetch from cloud data. Trim to start of the read range - this is
-        // incorrect because clients can still see earlier offsets but will work
-        // if they won't use aborted ranges from this request to filter batches
-        // belonging to earlier offsets.
-        trim_at = base_rp;
-    }
-
-    std::vector<cluster::rm_stm::tx_range> target;
-    target.reserve(source.size());
-    for (const auto& range : source) {
-        target.push_back(cluster::rm_stm::tx_range{
-          .pid = range.pid,
-          .first = ot_state->from_log_offset(std::max(trim_at, range.first)),
-          .last = ot_state->from_log_offset(range.last)});
-    }
-
+    cloud_storage::offset_range offsets = {
+      .begin = base,
+      .end = last,
+      .begin_rp = base_rp,
+      .end_rp = last_rp,
+    };
+    auto target = co_await _partition->aborted_transactions_cloud(offsets);
```

Comment:
I'm concerned that fetch.cc relies on partition_proxy to execute:
 -  `make_reader` to fetch data
 -  `aborted_transactions` to provide aborted for the client  to filter out

If we look into `make_reader` it has two branches: one is going via `partition::make_cloud_reader` and another via `partition::make_reader` while `aborted_transactions` always goes via `partition::aborted_transactions_cloud` so it seems we may have a situation when `fetch.cc` uses `partition::make_reader` to read data and then uses `partition::aborted_transactions_cloud` to get the list of aborted transactions.

It looks like if there is synchronization issues (can there be sync issues? I'm not competent enough with SI code to rule them out) we may end up with a situation when not all aborted transactions are returned and a client will treat aborted transactions if they were committed.

Reply:
yes, I need to keep old branch and add a new one for SI

Reply:
I updated the PR. Now it checks local snapshot first. If local snapshot doesn't have the required offset range it uses metadata from S3.